src/persist-client/src/cli/admin.rs

```diff
@@ -125,10 +134,115 @@ pub async fn run(command: AdminArgs) -> Result<(), anyhow::Error> {
             .await?;
             info_log_non_zero_metrics(&metrics_registry.gather());
         }
+        Command::RestoreBlob(args) => {
+            let StoreArgs {
+                consensus_uri,
+                blob_uri,
+            } = args;
+            let commit = command.commit;
+            let cfg = PersistConfig::new(&BUILD_INFO, SYSTEM_TIME.clone());
+            let metrics_registry = MetricsRegistry::new();
+            let metrics = Arc::new(Metrics::new(&cfg, &metrics_registry));
+            let consensus =
+                make_consensus(&cfg, &consensus_uri, commit, Arc::clone(&metrics)).await?;
+            let blob = make_blob(&cfg, &blob_uri, commit, Arc::clone(&metrics)).await?;
+            let versions = StateVersions::new(
+                cfg.clone(),
+                Arc::clone(&consensus),
+                Arc::clone(&blob),
+                metrics,
+            );
+
+            let mut shards = consensus.list_keys();
+            let mut not_restored = vec![];
+            while let Some(shard) = shards.next().await {
+                let shard_id = shard?;
+                let shard_id = ShardId::from_str(&shard_id).expect("invalid shard id");
+
+                not_restored.extend(
+                    restore_blob(&versions, blob.as_ref(), &cfg.build_version, shard_id).await?,
+                );
+            }
+            if !not_restored.is_empty() {
+                bail!("referenced blobs were not restored: {not_restored:#?}")
+            }
+        }
     }
     Ok(())
 }
 
+/// Attempt to restore all the blobs referenced by the current state in consensus.
+/// Returns a list of blobs that were not possible to restore.
+async fn restore_blob(
```

Comment:
would be nice to have some basic unit test coverage of this (the MemBlob comment I left will make this easier). kind just thinking: start up persist, write some data, delete blobs out from under it, restore them, and then read the data. I _think_ that won't tickle any of the correctness issues that we'd expect in an online restore?